File: function-workshop/import_zip.mjs
#!/usr/bin/env node
/**
 * Import a zip package into a function-workshop workfolder.
 *
 * Unpacks functions from a zip (created by the frontend export or export_zip.mjs)
 * into a workfolder directory and reconstructs registry.json.
 *
 * Usage:
 *   node import_zip.mjs --workfolder workfolders/my-suite path/to/package.zip
 *   node import_zip.mjs --list path/to/package.zip
 */

import { readFileSync, writeFileSync, existsSync, mkdirSync } from 'fs';
import { resolve, basename } from 'path';
import { createRequire } from 'module';

const require = createRequire(import.meta.url);
const JSZip = require('jszip');

// ── XML dependency parsing ────────────────────────────────────────────

function parseCustomFunctionsFromXml(xmlContent) {
  const deps = [];
  // Match <Function name="..."> inside <CustomFunctions>
  const cfMatch = xmlContent.match(/<CustomFunctions>([\s\S]*?)<\/CustomFunctions>/);
  if (!cfMatch) return deps;

  const funcRegex = /<Function\s[^>]*name="([^"]*)"[^>]*>/g;
  let m;
  while ((m = funcRegex.exec(cfMatch[1])) !== null) {
    if (m[1]) deps.push(m[1].toUpperCase());
  }
  return deps;
}

// ── Sheet type detection ──────────────────────────────────────────────

function detectSheetType(xmlContent) {
  const match = xmlContent.match(/sheetType="([^"]*)"/);
  return match ? match[1] : 'standard';
}

// ── Read manifest entries (v1 + v2 formats) ───────────────────────────

function readManifestEntries(manifest) {
  const entries = [];

  // v1 format: manifest.functions
  const functions = manifest.functions || {};
  if (Object.keys(functions).length > 0) {
    for (const [uid, meta] of Object.entries(functions)) {
      entries.push({
        uid,
        name: (meta.name || '').toUpperCase(),
        description: meta.description || '',
        sheetType: meta.sheetType || 'standard',
      });
    }
    return entries;
  }

  // v2 format: manifest.sheets
  const sheets = manifest.sheets || {};
  for (const [sheetId, meta] of Object.entries(sheets)) {
    entries.push({
      uid: sheetId,
      functionId: meta.functionId || null,
      name: (meta.name || '').toUpperCase(),
      description: meta.description || '',
      sheetType: meta.type || 'standard',
    });
  }

  // Also check manifest.spreadsheets (v1 without functions)
  if (entries.length === 0) {
    const spreadsheets = manifest.spreadsheets || {};
    for (const [sheetId, meta] of Object.entries(spreadsheets)) {
      entries.push({
        uid: sheetId,
        name: (meta.name || '').toUpperCase(),
        description: meta.description || '',
        sheetType: meta.type || 'standard',
      });
    }
  }

  return entries;
}

// ── List zip contents ─────────────────────────────────────────────────

async function listZipContents(zipPath) {
  const zipData = readFileSync(zipPath);
  const zip = await JSZip.loadAsync(zipData);

  const manifestFile = zip.file('manifest.json');
  if (!manifestFile) {
    console.error('Error: Zip has no manifest.json');
    process.exit(1);
  }

  const manifest = JSON.parse(await manifestFile.async('string'));
  const entries = readManifestEntries(manifest);

  if (entries.length === 0) {
    console.log('(empty package)');
    return;
  }

  const nameWidth = Math.max(...entries.map(e => e.name.length));
  const typeWidth = Math.max(...entries.map(e => e.sheetType.length));

  for (const entry of entries.sort((a, b) => a.name.localeCompare(b.name))) {
    const name = entry.name.padEnd(nameWidth);
    const stype = entry.sheetType.padEnd(typeWidth);
    const desc = entry.description;
    if (desc) {
      console.log(`  ${name}  ${stype}  ${desc}`);
    } else {
      console.log(`  ${name}  ${stype}`);
    }
  }
}

// ── Import zip into workfolder ────────────────────────────────────────

async function importZip(workfolderDir, zipPath) {
  if (!existsSync(zipPath)) {
    console.error(`Error: Zip not found: ${zipPath}`);
    process.exit(1);
  }

  mkdirSync(workfolderDir, { recursive: true });

  // Load existing registry (if any)
  const regPath = resolve(workfolderDir, 'registry.json');
  let registry = {};
  if (existsSync(regPath)) {
    registry = JSON.parse(readFileSync(regPath, 'utf-8'));
  }

  const zipData = readFileSync(zipPath);
  const zip = await JSZip.loadAsync(zipData);

  const manifestFile = zip.file('manifest.json');
  if (!manifestFile) {
    console.error('Error: Zip has no manifest.json');
    process.exit(1);
  }

  const manifest = JSON.parse(await manifestFile.async('string'));
  const entries = readManifestEntries(manifest);
  let imported = 0;

  for (const entry of entries) {
    const uid = entry.uid;
    const funcName = entry.name;

    // Try v1 path first, then v2
    let xmlContent = null;
    for (const xmlZipPath of [`functions/${uid}.xml`, `sheets/${uid}.xml`, `sheets/${uid}.published.xml`, `spreadsheets/${uid}.xml`]) {
      const file = zip.file(xmlZipPath);
      if (file) {
        xmlContent = await file.async('string');
        break;
      }
    }

    if (xmlContent === null) {
      console.error(`Warning: No XML found for ${funcName}, skipping`);
      continue;
    }

    // Extract JS (may not exist)
    let jsContent = null;
    for (const jsZipPath of [`functions/${uid}.js`, `sheets/${uid}.published.js`]) {
      const file = zip.file(jsZipPath);
      if (file) {
        jsContent = await file.async('string');
        break;
      }
    }

    // Write to workfolder directory (sanitize name to prevent path traversal)
    const safeName = basename(funcName);
    const xmlFilename = `${safeName}.xml`;
    const jsFilename = `${safeName}.js`;

    writeFileSync(resolve(workfolderDir, xmlFilename), xmlContent);

    if (jsContent) {
      writeFileSync(resolve(workfolderDir, jsFilename), jsContent);
    }

    // Detect dependencies from XML
    const depNames = parseCustomFunctionsFromXml(xmlContent);

    // Detect sheet type
    let sheetType = entry.sheetType;
    if (sheetType === 'spreadsheet') sheetType = 'standard';

    // Build registry entry
    // Use functionId (from manifest) when available — this is the ID that
    // other XMLs reference in <CustomFunctions id="...">, so the transpiler
    // can resolve DAGs by the same key. Fall back to sheet ID for unpublished
    // sheets (display sheets, drafts) that have no functionId.
    const regEntry = {
      uuid: entry.functionId || uid,
      xml: xmlFilename,
      js: jsFilename,
      sheetType,
      dependencies: depNames,
    };
    if (entry.description) regEntry.description = entry.description;

    registry[funcName] = regEntry;

    const status = jsContent ? 'xml+js' : 'xml only';
    console.log(`  Imported ${funcName} (${status})`);
    imported++;
  }

  // Save registry
  writeFileSync(regPath, JSON.stringify(registry, null, 2) + '\n');

  console.log(`\nImported ${imported} function(s) into ${workfolderDir}`);

  const hasNoJs = Object.values(registry).some(
    r => r.js && !existsSync(resolve(workfolderDir, r.js))
  );
  if (hasNoJs) {
    console.log(`Note: Some functions have no JS — run \`node transpile.mjs --workfolder ${workfolderDir} --all\` to transpile`);
  }
}

// ── CLI ───────────────────────────────────────────────────────────────

function parseArgs() {
  const args = process.argv.slice(2);
  const opts = { workfolder: null, list: false, zipfile: null };

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--workfolder' && args[i + 1]) {
      opts.workfolder = args[++i];
    } else if (args[i] === '--list') {
      opts.list = true;
    } else if (!args[i].startsWith('-')) {
      opts.zipfile = args[i];
    }
  }

  return opts;
}

async function main() {
  const opts = parseArgs();

  if (!opts.zipfile) {
    console.error('Error: Zip file path is required');
    process.exit(1);
  }

  if (opts.list) {
    await listZipContents(opts.zipfile);
  } else if (opts.workfolder) {
    await importZip(opts.workfolder, opts.zipfile);
  } else {
    console.error('Error: Either --list or --workfolder is required');
    process.exit(1);
  }
}

main();

File: function-workshop/import_zip.py
#!/usr/bin/env python3
"""
Import a zip package into a function-workshop workfolder.

Unpacks functions from a zip (created by the frontend export or export_zip.py)
into a workfolder directory and reconstructs registry.json.

Usage:
    python import_zip.py --workfolder workfolders/my-suite path/to/package.zip
    python import_zip.py --list path/to/package.zip
"""

import argparse
import json
import os
import sys
import zipfile
import xml.etree.ElementTree as ET


def parse_custom_functions_from_xml(xml_content):
    """Extract CustomFunctions dependencies from XML content."""
    try:
        root = ET.fromstring(xml_content)
    except ET.ParseError:
        return []

    cf = root.find("CustomFunctions")
    if cf is None:
        return []

    deps = []
    for func_elem in cf.findall("Function"):
        name = func_elem.get("name")
        if name:
            deps.append(name.upper())
    return deps


def detect_sheet_type(xml_content):
    """Detect sheet type from XML root element."""
    try:
        root = ET.fromstring(xml_content)
        return root.get("sheetType", "standard")
    except ET.ParseError:
        return "standard"


def read_manifest_entries(zf):
    """Read function entries from a zip manifest, handling both v1 and v2 formats.

    Returns list of dicts with keys: uid, name, description, sheetType.
    """
    try:
        manifest = json.loads(zf.read("manifest.json"))
    except KeyError:
        print("Error: Zip has no manifest.json", file=sys.stderr)
        sys.exit(1)

    entries = []

    # v1 format: manifest.functions
    functions = manifest.get("functions", {})
    if functions:
        for uid, meta in functions.items():
            entries.append({
                "uid": uid,
                "name": meta.get("name", "").upper(),
                "description": meta.get("description", ""),
                "sheetType": meta.get("sheetType", "standard"),
            })
        return entries, manifest

    # v2 format: manifest.sheets
    sheets = manifest.get("sheets", {})
    for sheet_id, meta in sheets.items():
        entries.append({
            "uid": sheet_id,
            "name": meta.get("name", "").upper(),
            "description": meta.get("description", ""),
            "sheetType": meta.get("type", "standard"),
        })

    # Also check manifest.spreadsheets (v1 without functions)
    if not entries:
        spreadsheets = manifest.get("spreadsheets", {})
        for sheet_id, meta in spreadsheets.items():
            entries.append({
                "uid": sheet_id,
                "name": meta.get("name", "").upper(),
                "description": meta.get("description", ""),
                "sheetType": meta.get("type", "standard"),
            })

    return entries, manifest


def list_zip_contents(zip_path):
    """List functions in a zip without importing."""
    with zipfile.ZipFile(zip_path) as zf:
        entries, _ = read_manifest_entries(zf)

    if not entries:
        print("(empty package)")
        return

    # Calculate column widths
    name_width = max(len(e["name"]) for e in entries)
    type_width = max(len(e["sheetType"]) for e in entries)

    for entry in sorted(entries, key=lambda e: e["name"]):
        name = entry["name"].ljust(name_width)
        stype = entry["sheetType"].ljust(type_width)
        desc = entry["description"]
        if desc:
            print(f"  {name}  {stype}  {desc}")
        else:
            print(f"  {name}  {stype}")


def import_zip(workfolder_dir, zip_path):
    """Import a zip into a workfolder directory."""
    if not os.path.exists(zip_path):
        print(f"Error: Zip not found: {zip_path}", file=sys.stderr)
        sys.exit(1)

    # Create workfolder directory
    os.makedirs(workfolder_dir, exist_ok=True)

    # Load existing registry (if any)
    reg_path = os.path.join(workfolder_dir, "registry.json")
    if os.path.exists(reg_path):
        with open(reg_path) as f:
            registry = json.load(f)
    else:
        registry = {}

    with zipfile.ZipFile(zip_path) as zf:
        entries, manifest = read_manifest_entries(zf)
        imported = 0

        for entry in entries:
            uid = entry["uid"]
            func_name = entry["name"]

            # Try v1 path first, then v2
            xml_content = None
            for xml_zip_path in [f"functions/{uid}.xml", f"sheets/{uid}.xml", f"spreadsheets/{uid}.xml"]:
                try:
                    xml_content = zf.read(xml_zip_path).decode("utf-8")
                    break
                except KeyError:
                    continue

            if xml_content is None:
                print(f"Warning: No XML found for {func_name}, skipping", file=sys.stderr)
                continue

            # Extract JS (may not exist)
            js_content = None
            for js_zip_path in [f"functions/{uid}.js", f"sheets/{uid}.published.js"]:
                try:
                    js_content = zf.read(js_zip_path).decode("utf-8")
                    break
                except KeyError:
                    continue

            # Write to workfolder directory (sanitize name to prevent path traversal)
            safe_name = os.path.basename(func_name)
            xml_filename = f"{safe_name}.xml"
            js_filename = f"{safe_name}.js"

            with open(os.path.join(workfolder_dir, xml_filename), "w") as f:
                f.write(xml_content)

            if js_content:
                with open(os.path.join(workfolder_dir, js_filename), "w") as f:
                    f.write(js_content)

            # Detect dependencies from XML
            dep_names = parse_custom_functions_from_xml(xml_content)

            # Detect sheet type
            sheet_type = entry["sheetType"]
            if sheet_type == "spreadsheet":
                sheet_type = "standard"

            # Build registry entry
            reg_entry = {
                "uuid": uid,
                "xml": xml_filename,
                "js": js_filename,
                "sheetType": sheet_type,
                "dependencies": dep_names,
            }
            if entry["description"]:
                reg_entry["description"] = entry["description"]

            registry[func_name] = reg_entry

            status = "xml+js" if js_content else "xml only"
            print(f"  Imported {func_name} ({status})")
            imported += 1

    # Save registry
    with open(reg_path, "w") as f:
        json.dump(registry, f, indent=2)
        f.write("\n")

    print(f"\nImported {imported} function(s) into {workfolder_dir}")
    if any(not os.path.exists(os.path.join(workfolder_dir, r["js"])) for r in registry.values()):
        print("Note: Some functions have no JS — run `python transpile.py --workfolder {} --all` to transpile".format(workfolder_dir))


def main():
    parser = argparse.ArgumentParser(description="Import zip into workfolder directory")
    parser.add_argument("--workfolder", help="Path to workfolder directory")
    parser.add_argument("--list", action="store_true", help="List zip contents without importing")
    parser.add_argument("zipfile", help="Path to zip file to import")

    args = parser.parse_args()

    if args.list:
        list_zip_contents(args.zipfile)
    elif args.workfolder:
        import_zip(args.workfolder, args.zipfile)
    else:
        print("Error: Either --list or --workfolder is required", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
